cpp/src/strings/extract.cu

```diff
@@ -0,0 +1,138 @@
+    {
+        u_char data1[stack_size], data2[stack_size];
+        prog.set_stack_mem(data1,data2);
+        if( d_strings.is_null(idx) )
+            return string_index_pair{nullptr,0};
+        string_view d_str = d_strings.element<string_view>(idx);
+        string_index_pair result{nullptr,0};
+        int32_t begin = 0, end = d_str.length();
+        int rtn = prog.find(idx,d_str,begin,end);
+        if( rtn > 0 )
+            rtn = prog.extract(idx,d_str,begin,end,column_index);
+        if( rtn > 0 )
+        {
+            auto offset = d_str.byte_offset(begin);
+            result = string_index_pair{ d_str.data() + offset, d_str.byte_offset(end)-offset };
+        }
+        return result;
+    }
+};
+
+} // namespace
+
+//
+std::unique_ptr<experimental::table> extract( strings_column_view const& strings,
+                                              std::string const& pattern,
+                                              rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource(),
+                                              cudaStream_t stream = 0)
+{
+    auto strings_count = strings.size();
+    auto strings_column = column_device_view::create(strings.parent(),stream);
+    auto d_strings = *strings_column;
+
+    // compile regex into device object
+    auto prog = Reprog_device::create(pattern,get_character_flags_table(),strings_count,stream);
+    auto d_prog = *prog;
+    // extract should include groups
+    int groups = d_prog.group_counts();
+    CUDF_EXPECTS( groups > 0, "Group indicators not found in regex pattern");
+
+    // build a result column for each group
+    std::vector<std::unique_ptr<column>> results;
+    auto execpol = rmm::exec_policy(stream);
+    auto regex_insts = d_prog.insts_counts();
+    for( int32_t column_index=0; column_index < groups; ++column_index )
+    {
+        rmm::device_vector<string_index_pair> indices(strings_count);
+        string_index_pair* d_indices = indices.data().get();
+
+        if( (regex_insts > MAX_STACK_INSTS) || (regex_insts <= RX_SMALL_INSTS) )
+            thrust::transform(execpol->on(stream),
+                thrust::make_counting_iterator<size_type>(0),
+                thrust::make_counting_iterator<size_type>(strings_count),
+                d_indices, extract_fn<RX_STACK_SMALL>{d_prog, d_strings, column_index});
+        else if( regex_insts <= RX_MEDIUM_INSTS )
+            thrust::transform(execpol->on(stream),
+                thrust::make_counting_iterator<size_type>(0),
+                thrust::make_counting_iterator<size_type>(strings_count),
+                d_indices, extract_fn<RX_STACK_MEDIUM>{d_prog, d_strings, column_index});
+        else
+            thrust::transform(execpol->on(stream),
+                thrust::make_counting_iterator<size_type>(0),
+                thrust::make_counting_iterator<size_type>(strings_count),
+                d_indices, extract_fn<RX_STACK_LARGE>{d_prog, d_strings, column_index});
```

Comment:
Same comment as above, also you can put the `determine_stack_size` function outside of the `for` loop.